src/DynamoCore/Models/RecordableCommands.cs

```diff
@@ -232,6 +287,69 @@ internal static RecordableCommand Deserialize(XmlElement element)
             #endregion
         }
 
+        /// <summary>
+        /// This class is base for those RecordableCommands that should have 
+        /// Guid NodeId that causes the problems during deserialization
+        /// </summary>
+        [DataContract]
+        public abstract class NodeSpecificRecordableCommand : RecordableCommand
+        {
+            internal Guid NodeId { get; private set; }
+
+            [DataMember]
+            internal string NodeIdAsString
+            {
+                get { return NodeId.ToString(); }
+                private set
+                {
+                    Guid guid;
+                    NodeId = Guid.TryParse(value, out guid) ? guid : new Guid();
+                }
+            }
+
+            protected NodeSpecificRecordableCommand(string node)
+            {
+                NodeIdAsString = node;
+            }
+
+            protected NodeSpecificRecordableCommand(Guid node)
+            {
+                NodeId = node;
+            }
+        }
+
+        /// <summary>
+        /// This class is base for those RecordableCommands that should have 
+        /// Guid ModelGuid that causes the problems during deserialization
+        /// </summary>
+        [DataContract]
+        public abstract class ModelSpecificRecordableCommand : RecordableCommand
+        {
+            public Guid ModelGuid { get; private set; }
+
+            [DataMember]
+            public string ModelGuidAsString
+            {
+                get { return ModelGuid.ToString(); }
+                private set
+                {
+                    Guid guid;
+                    ModelGuid = Guid.TryParse(value, out guid) ? guid : new Guid();
+                }
+            }
+
+            protected ModelSpecificRecordableCommand(string node)
+            {
+                ModelGuidAsString = node;
+            }
+
+            protected ModelSpecificRecordableCommand(Guid node)
+            {
+                ModelGuid = node;
+            }
+        }
+
```

Comment:
The way `ModelSpecificRecordableCommand` is introduced now has a few issues:

1. The class contains `ModelGuid` but it's only used in `ModelSpecificRecordableCommand` constructor. The class should be responsible for serializing it, but it is not.

2. The job of serializing `ModelGuid` is left to the derived classes, this can be easily overlooked.

3. It is a base class for `UpdateModelValueCommand`, which is meant to deal with multiple `ModelBase` at the same time. This relationship does not make too much sense.

4. `ModelSpecificRecordableCommand` is nearly empty, its only function is to initialize `ModelGuid` data member.

5. It is completely identical to `NodeSpecificRecordableCommand`, there's no clear distinction between the two that warrant having another class.


Here's what I would suggest:

1. Merge `ModelSpecificRecordableCommand` and `NodeSpecificRecordableCommand` into one abstract class called `ModelBasedRecordableCommand`.

2. Add only a single constructor to `ModelBasedRecordableCommand` that takes `IEnumerable<Guid> modelGuids`. This class is designed to support multiple `ModelBase` internally, but derived classes can choose to work with only one.

3. Make constructors of all `RecordableCommand` derived classes take `XmlElement element`, and `ModelBasedRecordableCommand` should deserialize `ModelGuids` list from `XmlElement`.

4. Retain backward compatibilities by reading a single `ModelGuid` from `XmlElement` in `ModelBasedRecordableCommand`, but serialization should write out multiple entries from `ModelGuids` into the `XmlElement`.

5. Make sure all derived constructors and `SerializeCore` methods call `base` methods accordingly.

6. Increase the existing test cases to cover the new changes (add directly in `RecordedTests` class).

This does not have to affect this PR, I have created [an internally tracked task](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6572) for you to work on right after this :)